### PR TITLE
fix: fix incorect coverage reporting (fix #375)

### LINKED ISSUE #375: Wrong coverage report
### Describe the bug

The coverage doesn't make sense with the actual code

image
image
image

### Reproduction

Github repo: https://github.com/Nisgrak/vitest-coverage-bug

Execute `pnpm i && pnpm test:unit`

### System Info

```shell
System:
    OS: Linux 5.10 Ubuntu 20.04.3 LTS (Focal Fossa)
    CPU: (16) x64 AMD Ryzen 7 4800H with Radeon Graphics
    Memory: 1.24 GB / 24.56 GB
    Container: Yes
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.13.1 - ~/.volta/tools/image/node/16.13.1/bin/node
    Yarn: 1.22.17 - ~/.volta/tools/image/yarn/1.22.17/bin/yarn
    npm: 8.1.2 - ~/.volta/tools/image/node/16.13.1/bin/npm
  Browsers:
    Firefox: 95.0.1
  npmPackages:
    vite: ^2.7.2 => 2.7.10 
    vitest: ^0.0.124 => 0.0.124
```

### Used Package Manager

pnpm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by Demivan**
This reverts #584 - runtime coverage is not reporting functions that are not used as not covered.

Added manual soucemap adjustments because Vitest is prepending code to every file. 

Fixes #375

image

### CODE REVIEW

**Review by mitchelvanbever — COMMENTED** [review]
Tested it agains a local branch but playing around with my test files did not result in different coverage and there is one line always uncovered no matter what test I write to accomodate for covering that line.

**Comment by Demivan** [comment]
@mitchelvanbever can you share a repo?

**Comment by mitchelvanbever** [comment]
> @mitchelvanbever can you share a repo?

Just tried with the example folder react-testing-lib and it shows the same behaviour. The other node project is private, but I can try it with another public repo I have

**Comment by Demivan** [comment]
This change fixed coverage for `coverage-test` test. And tested coverage with `fluent-vue` - it is correct now.

**Comment by Demivan** [comment]
> Just tried with the example folder react-testing-lib and it shows the same behaviour. The other node project is private, but I can try it with another public repo I have

Tested with react-testing-lib example after skipping error test, error branch is now not covered:

image

**Comment by mitchelvanbever** [comment]
> > Just tried with the example folder react-testing-lib and it shows the same behaviour. The other node project is private, but I can try it with another public repo I have
> 
> Tested with react-testing-lib example after skipping error test, error branch is now not covered:
> 
> image

I feel really stupid 🙈  I think I just didn't configure the coverage properly yesterday when trying it out. I was too excited about it 😬 I tried again on the example dir, with the proper configuration and I can confirm to have the same result as you show.

**Review by antfu — APPROVED** [review]

**Comment by antfu** [comment]
Epic, awesome work @Demivan!

File: packages/vitest/src/integrations/coverage.ts [hunk 1]
```diff
@@ -36,47 +39,79 @@ export function resolveC8Options(options: C8Options, root: string): ResolvedC8Op
...
+          ...map,
+          sources,
+        },
+      }
+    }))
+
+  // This is a magic number. It corresponds to the amount of code
+  // that we add in packages/vite-node/src/client.ts:114 (vm.runInThisContext)
+  // TODO: Include our transformations in soucemaps
+  const offset = 190
```
Comment (antfu) [inline packages/vitest/src/integrations/coverage.ts:100]: Later we should calculate this number based on the length of the arguments, but I am fine to have the magic number for now.

**Review by sheremet-va — APPROVED** [review]

**Comment by cexbrayat** [comment]
FYI, the introduction of `takeCoverage` makes it incompatible with node < v14.18
As Stackblitz uses v14.16, the runnable examples can no longer launch the code coverage in WebContainers:

```
❯ npm run coverage
$ vitest run --coverage
(node:56) UnhandledPromiseRejectionWarning: file:///home/projects/vitest-dev-vitest-ilb2yu/node_modules/vitest/dist/create-a1bc541a.js:13
import { takeCoverage } from 'v8';
         ^^^^^^^^^^^^
```

**Comment by Demivan** [comment]
@cexbrayat Yeah, we have issue about this already #670. Not sure what is the best solution. Without `takeCoverage` coverage will be incomplete.

**Pull request merged**